Mathlib/AlgebraicGeometry/EllipticCurve/Reduction.lean

```diff
@@ -0,0 +1,229 @@
+/-
+Copyright (c) 2025 Bryan Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bryan Wang
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.VariableChange
+import Mathlib.RingTheory.DedekindDomain.AdicValuation
+import Mathlib.RingTheory.Valuation.Discrete.Basic
+
+/-!
+# Reduction of Weierstrass curves over local fields
+
+This file defines reduction of Weierstrass curves over local fields, or more generally,
+fraction fields of discrete valuation rings.
+
+## Main definitions
+
+* `IsIntegral`: a predicate expressing that a given Weierstrass equation
+  has integral coefficients.
+* `IsMinimal`: a predicate expressing that a given Weierstrass equation
+  has minimal valuation of discriminant among all isomorphic integral Weierstrass equations.
+* `reduction`: the reduction of a Weierstrass curve given by a minimal Weierstrass equation,
+  which is a Weierstrass curve over the residue field.
+* `IsGoodReduction`: a predicate expressing that a given minimal Weierstrass equation
+  has valuation of its discriminant equal to zero.
+
+## Main statements
+
+* `exists_integral`: any Weierstrass curve is isomorphic to one given by
+  an integral Weierstrass equation.
+* `exists_minimal`: any Weierstrass curve is isomorphic to one given by
+  a minimal Weierstrass equation.
+
+## References
+
+* [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, minimal weierstrass equation, reduction
+-/
+
+namespace WeierstrassCurve
+
+variable (R : Type*) [CommRing R] [IsDomain R] [IsDiscreteValuationRing R]
+variable {K : Type*} [Field K] [Algebra R K] [IsFractionRing R K]
+
+section Minimal
+
+open IsDedekindDomain.HeightOneSpectrum
+open IsDiscreteValuationRing
+
+/-- A Weierstrass equation over the fraction field `K` is integral if
+it has coefficients in the DVR `R`. -/
+@[mk_iff]
+class IsIntegral (W : WeierstrassCurve K) : Prop where
+  integral : ∃ W_int : WeierstrassCurve R, W = W_int.baseChange K
+
+lemma isIntegral_of_val_le_one {W : WeierstrassCurve K}
+    (h₁ : valuation K (maximalIdeal R) W.a₁ ≤ 1)
```

Comment:
Have you considered using `ValuationRing.valuation` here and elsewhere in this file? I think in principle this would allow for more generalization later on.